common/lib/xmodule/xmodule/js/src/capa/display.js

```diff
@@ -1157,32 +1160,61 @@
          */
         Problem.prototype.disableAllButtonsWhileRunning = function(operationCallback, isFromCheckOperation) {
             var that = this;
-            this.enableAllButtons(false, isFromCheckOperation);
+            var allButtons = [this.resetButton, this.saveButton, this.showButton, this.hintButton, this.submitButton];
+            var disabledButtons = this.disableButtons(allButtons, isFromCheckOperation);
             return operationCallback().always(function() {
-                return that.enableAllButtons(true, isFromCheckOperation);
+                return that.enableButtons(disabledButtons, isFromCheckOperation);
             });
         };
 
         /**
-         * Used to enable/disable all buttons in problem.
+         * Disables all buttons by adding the disabled attribute. The submit button is checked separately due to
+         *     the changing text it contains.
          *
          * params:
-         *     'enable' is a boolean to determine enabling/disabling of buttons.
-         *     'isFromCheckOperation' is a boolean to keep track if operation was initiated
+         *     'changeText' is a boolean to keep track if operation was initiated
          *         from submit so that text of submit button will also be changed while disabling/enabling
          *         the submit button.
+         *
+         * returns:
+         *      an array of the buttons that have had their state modified to disabled.
+         *          This return value should be preserved and used to re-enable them when input should be accepted again
          */
-        Problem.prototype.enableAllButtons = function(enable, isFromCheckOperation) {
-            // Called by disableAllButtonsWhileRunning to automatically disable all buttons while check,reset, or
-            // save internal are running. Then enable all the buttons again after it is done.
-            if (enable) {
-                this.resetButton.add(this.saveButton).add(this.hintButton).add(this.showButton).
-                    removeAttr('disabled');
-            } else {
-                this.resetButton.add(this.saveButton).add(this.hintButton).add(this.showButton).
-                    attr({disabled: 'disabled'});
-            }
-            return this.enableSubmitButton(enable, isFromCheckOperation);
+        Problem.prototype.disableButtons = function(buttons, changeText) {
+            var that = this;
+            var disabledButtons = [];
+            buttons.forEach(function(button) {
+                if (!button.attr('disabled')) {
+                    disabledButtons.push(button);
+                    if (button === that.submitButton) {
+                        that.enableSubmitButton(false, changeText);
+                    } else {
+                        button.attr({disabled: 'disabled'});
+                    }
+                }
+            });
+            return disabledButtons;
```

Comment:
I'd like disableButtons and enableButtons to be symmetrical. disableButtons checks that the button is disabled before it does anything, but enableButtons doesn't check that it is enabled. disableButtons also returns something, whereas enableButtons doesn't return the array.

My preference would be to have both methods follow the pattern of enableButtons (assuming there is no harm in disabling something that is already disabled). And then there is no need to return the array of disabled buttons, since the caller already has the array they have called the method with.